src/tools/rbd/action/Import.cc

```diff
@@ -63,9 +295,219 @@ class C_Import : public Context {
   uint64_t m_offset;
 };
 
+static int do_import_header(int fd, int import_format, uint64_t &size, librbd::ImageOptions& opts)
+{
+  int r = 0;
+  char image_buf[utils::RBD_IMAGE_BANNER_V2.size() + 1];
+
+  // There is no header in v1 image.
+  if (import_format == 1)
+    return r;
+
+  if (fd == 0 || size < utils::RBD_IMAGE_BANNER_V2.size()) {
+    r = -EINVAL;
+    return r;
+  }
+  r = safe_read_exact(fd, image_buf, utils::RBD_IMAGE_BANNER_V2.size());
+  if (r < 0)
+    return r;
+
+  image_buf[utils::RBD_IMAGE_BANNER_V2.size()] = '\0';
+  if (strcmp(image_buf, utils::RBD_IMAGE_BANNER_V2.c_str())) {
+    // Old format
+    r = -EINVAL;
+    return r;
+  }
+
+  r = safe_read_exact(fd, &size, 8);
+  if (r < 0) {
+    return r;
+  }
+
+  while (1) {
+    __u8 tag;
+    uint64_t length;
+    r = safe_read_exact(fd, &tag, 1);
+    if (r < 0) {
+      return r;
+    }
+
+    if (tag == RBD_EXPORT_IMAGE_END) {
+      break;
+    } else {
+      r = safe_read_exact(fd, &length, 8);
+      if (r < 0) {
+	return r;
+      }
+
+      if (tag == RBD_EXPORT_IMAGE_ORDER) {
+	uint64_t order = 0;
+	r = safe_read_exact(fd, &order, 8);
+	if (r < 0) {
+	  return r;
+	}
+	if (opts.get(RBD_IMAGE_OPTION_ORDER, &order) != 0) {
+	  opts.set(RBD_IMAGE_OPTION_ORDER, order);
+	}
+      } else if (tag == RBD_EXPORT_IMAGE_FORMAT) {
+	uint64_t format = 0;
+	r = safe_read_exact(fd, &format, 8);
+	if (r < 0) {
+	  return r;
+	}
+	if (opts.get(RBD_IMAGE_OPTION_FORMAT, &format) != 0) {
+	  opts.set(RBD_IMAGE_OPTION_FORMAT, format);
+	}
+      } else if (tag == RBD_EXPORT_IMAGE_FEATURES) {
+	uint64_t features = 0;
+	r = safe_read_exact(fd, &features, 8);
+	if (r < 0) {
+	  return r;
+	}
+	if (opts.get(RBD_IMAGE_OPTION_FEATURES, &features) != 0) {
+	  opts.set(RBD_IMAGE_OPTION_FEATURES, features);
+	}
+      } else if (tag == RBD_EXPORT_IMAGE_STRIPE_UNIT) {
+	uint64_t stripe_unit = 0;
+	r = safe_read_exact(fd, &stripe_unit, 8);
+	if (r < 0) {
+	  return r;
+	}
+	if (opts.get(RBD_IMAGE_OPTION_STRIPE_UNIT, &stripe_unit) != 0) {
+	  opts.set(RBD_IMAGE_OPTION_STRIPE_UNIT, stripe_unit);
+	}
+      } else if (tag == RBD_EXPORT_IMAGE_STRIPE_COUNT) {
+	uint64_t stripe_count = 0;
+	r = safe_read_exact(fd, &stripe_count, 8);
+	if (r < 0) {
+	  return r;
+	}
+	if (opts.get(RBD_IMAGE_OPTION_STRIPE_COUNT, &stripe_count) != 0) {
+	  opts.set(RBD_IMAGE_OPTION_STRIPE_COUNT, stripe_count);
+	}
+      } else {
+	std::cerr << "rbd: invalid tag in image properties zone: " << tag << "Skip it." << std::endl;
+	::lseek(fd, length, SEEK_CUR);
```

Comment:
Yes, but then you need to used decode too, i.e. read to a buffer using bufferlist::read_fd and call decode.